File: src/main/java/com/exactpro/th2/codec/json/TypeInfoSerializer.java
package com.exactpro.th2.codec.json;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.ser.std.StdSerializer;

import java.io.IOException;
import java.math.BigDecimal;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class TypeInfoSerializer extends StdSerializer<String> {
    private Pattern numberPattern = Pattern.compile("number\\((\\d*[.|,]\\d|\\d+)\\)");
    private Predicate<String> numberPredicate = numberPattern.asMatchPredicate();

    private Pattern booleanPattern = Pattern.compile("boolean\\((true|false)\\)");
    private Predicate<String> booleanPredicate = booleanPattern.asMatchPredicate();

    protected TypeInfoSerializer() {
        super(String.class);
    }

    @Override
    public void serialize(String value, JsonGenerator gen, SerializerProvider provider) throws IOException {
        if(numberPredicate.test(value)) {
            Matcher matcher = numberPattern.matcher(value);
            matcher.matches();
            gen.writeNumber(new BigDecimal(matcher.group(1)));
            return;
        }
        if(booleanPredicate.test(value)) {
            Matcher matcher = booleanPattern.matcher(value);
            matcher.matches();
            gen.writeBoolean(Boolean.parseBoolean(matcher.group(1)));
            return;
        }
        gen.writeString(value);
    }
}

File: src/main/kotlin/com/exactpro/th2/codec/json/TypeInfoSerializer.kt
/*
 * Copyright 2022-2022 Exactpro (Exactpro Systems Limited)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.exactpro.th2.codec.json

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.ser.std.StdSerializer
import java.io.IOException
import java.math.BigDecimal
import java.math.BigInteger
import kotlin.String
import kotlin.Throws


class TypeInfoSerializer: StdSerializer<String>(String::class.java) {
    @Throws(IOException::class)
    override fun serialize(value: String, gen: JsonGenerator, provider: SerializerProvider) = when {
        !value.endsWith(TYPE_SUFFIX) -> gen.writeString(value)
        value.startsWith(NUMBER_PREFIX) -> {
            value.substring(NUMBER_PREFIX_LENGTH, value.lastIndex).let { value ->
                if (value.any { it == '.' || it == ',' }) {
                    gen.writeNumber(BigDecimal(value))
                } else {
                    gen.writeNumber(BigInteger(value))
                }
            }
        }
        value.startsWith(BOOLEAN_PREFIX) -> {
            gen.writeBoolean(value.substring(BOOLEAN_PREFIX_LENGTH, value.lastIndex).toBoolean())
        }
        else -> gen.writeString(value)
    }

    companion object {
        private const val NUMBER_PREFIX = "number("
        private const val NUMBER_PREFIX_LENGTH = NUMBER_PREFIX.length
        private const val BOOLEAN_PREFIX = "boolean("
        private const val BOOLEAN_PREFIX_LENGTH = BOOLEAN_PREFIX.length
        private const val TYPE_SUFFIX = ")"
    }
}
